src/Workspaces/CoreTest/Editting/SyntaxEditorTests.cs

```diff
@@ -135,6 +137,190 @@ public class C
+            var cls = cu.Members[0];
+
+            var editor = GetEditor(cu);
+            var fieldX = editor.Generator.GetMembers(cls)[0];
+            var newFieldY = editor.Generator.FieldDeclaration("Y", editor.Generator.TypeExpression(SpecialType.System_String), Accessibility.Public);
+            insertAfterWithTracking(fieldX, newFieldY, editor);
+
+            var newRoot = editor.GetChangedRoot();
+            await VerifySyntaxAsync<CompilationUnitSyntax>(
+                newRoot,
+                @"
+public class C
+{
+    public int X;
+    public string Y;
+}");
+
+            var newFieldZ = editor.Generator.FieldDeclaration("Z", editor.Generator.TypeExpression(SpecialType.System_String), Accessibility.Public);
+            editor.ReplaceNode(newFieldY, newFieldZ);
+
+            newRoot = editor.GetChangedRoot();
+            await VerifySyntaxAsync<CompilationUnitSyntax>(
+                newRoot,
+                @"
+public class C
+{
+    public int X;
+    public string Z;
+}");
+        }
+
+        [Fact]
+        public async Task TestInsertBeforeWithTracking()
+        {
+            // InsertBefore overload #1
+            await TestInterBeforeWithTrackingCoreAsync((SyntaxNode node, SyntaxNode newNode, SyntaxEditor editor) =>
+            {
+                editor.InsertBefore(node, newNode);
+            });
+
+            // InsertBefore overload #2
+            await TestInterBeforeWithTrackingCoreAsync((SyntaxNode node, SyntaxNode newNode, SyntaxEditor editor) =>
+            {
+                editor.InsertBefore(node, new[] { newNode });
+            });
+        }
+
+        private async Task TestInterBeforeWithTrackingCoreAsync(Action<SyntaxNode, SyntaxNode, SyntaxEditor> insertBeforeWithTracking)
+        {
+            var code = @"
+public class C
+{
+    public int X;
+}";
+
+            var cu = SyntaxFactory.ParseCompilationUnit(code);
+            var cls = cu.Members[0];
+
+            var editor = GetEditor(cu);
+            var fieldX = editor.Generator.GetMembers(cls)[0];
+            var newFieldY = editor.Generator.FieldDeclaration("Y", editor.Generator.TypeExpression(SpecialType.System_String), Accessibility.Public);
+            insertBeforeWithTracking(fieldX, newFieldY, editor);
+
+            var newRoot = editor.GetChangedRoot();
+            await VerifySyntaxAsync<CompilationUnitSyntax>(
+                newRoot,
+                @"
+public class C
+{
+    public string Y;
+    public int X;
+}");
+
+            var newFieldZ = editor.Generator.FieldDeclaration("Z", editor.Generator.TypeExpression(SpecialType.System_String), Accessibility.Public);
+            editor.ReplaceNode(newFieldY, newFieldZ);
+
+            newRoot = editor.GetChangedRoot();
+            await VerifySyntaxAsync<CompilationUnitSyntax>(
+                newRoot,
+                @"
+public class C
+{
+    public string Z;
+    public int X;
+}");
+        }
+
         [Fact]
```

Comment:
are there tests that chain several updates to the same node in a row?  or tests taht do a replace of a larger tree, then a replace of a nested subtree, and then a replace of another nested subtree.  effectively, i want ot make sure that replacing works after several steps, not just one.

Reply:
can you have some additional tests that specifically do at least 3 replaces.  It would be good to do interesting mixes/matches of replaces both in the original tree and in the new trees added.   

Note: i imagine that the existing test really does prove this out.  But given the complexity here, i would just like a little bit more in teh way of testing just to get that additional bit of confidence.

Thanks!

Reply:
Sure, let me add bunch of additional tests. Thanks.